benchbuild/project.py

```diff
@@ -295,61 +281,80 @@
         finally:
             signals.handlers.deregister(fail_run_group)
 
-    def clean(self):
+    def clean(self) -> None:
         """Clean the project build directory."""
         builddir_p = local.path(self.builddir)
         builddir_p.delete()
 
-    def prepare(self):
-        """Prepare the build diretory."""
-        builddir_p = local.path(self.builddir)
-        if not builddir_p.exists():
-            builddir_p.mkdir()
+    def clone(self) -> 'Project':
+        """Create a deepcopy of ourself."""
+        new_p = copy.deepcopy(self)
+        new_p.run_uuid = uuid.uuid4()
+        return new_p
 
     @abstractmethod
-    def compile(self):
+    def compile(self) -> None:
         """Compile the project."""
 
-    def download(self, version=None):
-        """Auto-generated by with_* decorators."""
-        del version
-        LOG.error("Not implemented.")
+    @property
+    def id(self) -> str:
+        version_str = source.to_str(tuple(self.variant.values()))
+        return f"{self.name}/{self.group}/{version_str}/{self.run_uuid}"
 
-    def redirect(self):
+    def prepare(self) -> None:
+        """Prepare the build diretory."""
+        builddir_p = local.path(self.builddir)
+        if not builddir_p.exists():
+            builddir_p.mkdir()
+
+    def redirect(self) -> None:
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5694346)